tests/kernel/smp/src/main.c

```diff
@@ -72,20 +111,209 @@ void test_main(void)
+		zassert_true(tdata[i].executed == 1, "Threads didn't get"
+			     " scheduled on another core");
+	}
+}
+
+/**
+ * @brief Verify round robin scheduling on all cores
+ *
+ * @details The test verifies a scenario in which
+ * parent thread joins children to complete task in
+ * round robin.
+ */
+static void test_round_robin_all_cores(void)
+{
+	int i;
+
+	/* Make current thread preemptive */
+	k_thread_priority_set(k_current_get(), K_PRIO_PREEMPT(10));
+
+	k_sched_time_slice_set(TIME_SLICE_MS, 0);
+
+	/* Spawn threads of priority equal to
+	 * current thread */
+	spawn_threads(THREADS_NUM, K_PRIO_PREEMPT(10), 0);
+
+	k_busy_wait((DELAY_US * THREADS_NUM) /
+		    CONFIG_MP_NUM_CPUS);
+
+	for (i = 0; i < THREADS_NUM; i++) {
+		k_thread_abort(tdata[i].tid);
+	}
+
+	k_sched_time_slice_set(0, 0);
+
+	/* Test point: All THREADS_NUM spawned threads must have
+	 * executed in another core
+	 */
+	for (i = 0; i < THREADS_NUM; i++) {
+		zassert_true(tdata[i].executed == 1, "Threads didn't get"
+			     " scheduled on another core");
 	}
+}
+
+/**
+ * @brief Verify round robin on all cores with parent thread pending
+ *
+ * @details The test validates for the scenario where parent has
+ * to execute after all child processes execution is complete
+ */
+static void test_round_robin_all_cores_pend_parent(void)
+{
+	int i;
+
+	/* Make current thread preemptible */
+	k_thread_priority_set(k_current_get(),
+			      K_LOWEST_APPLICATION_THREAD_PRIO);
 
-	if (ok) {
-		TC_END_REPORT(TC_PASS);
-	} else {
-		TC_END_REPORT(TC_FAIL);
+	/* Set time slice */
+	k_sched_time_slice_set(TIME_SLICE_MS, 0);
+
+	/* Spawn threads of higher priority than current thread */
+	spawn_threads(THREADS_NUM,
+		      K_LOWEST_APPLICATION_THREAD_PRIO - 1, 0);
+
+	k_sleep(DELAY_US * (CONFIG_MP_NUM_CPUS - 1));
+
+	/* All threads must have completed execution */
+	for (i = 0; i < THREADS_NUM; i++) {
+		k_thread_abort(tdata[i].tid);
 	}
+
+	/* Reset time slice */
+	k_sched_time_slice_set(0, 0);
+
+	/* Test point: All THREADS_NUM spawned threads must have
+	 * executed.
+	 */
+	for (i = 0; i < THREADS_NUM; i++) {
+		zassert_true(tdata[i].executed == 1, "Threads didn't get"
+			     " scheduled on all the cores");
+	}
+}
+
+
+/**
+ * @brief Verify multi core processing of delayed threads
+ *
+ * @details The test verifies a scenario in which parent
+ * thread doesn't want itself to be preempted and children
+ * threads should execute in round robin with delayed start
+ */
+static void test_round_robin_delayed_threads(void)
+{
+	int i;
+
+	k_sched_time_slice_set(TIME_SLICE_MS, 0);
+
+	spawn_threads(THREADS_NUM, K_PRIO_PREEMPT(10),
+		      THREAD_START_DELAY_MS);
+
+	/* Threads shouldn't have executed as they are delayed*/
+	for (i = 0; i < THREADS_NUM; i++) {
+		zassert_false(tdata[i].executed == 1, "Delayed Threads"
+			      " didn't wait to execute");
+	}
+
+	k_busy_wait(((THREAD_START_DELAY_MS * 1000 + DELAY_US) *
+		     THREADS_NUM) / (CONFIG_MP_NUM_CPUS - 1));
```

Comment:
Same code structure as above again.  Also this looks like weirdly specific math for this timeout.  That's fragile.

If you want to spin on the current CPU waiting for completion, then it's probably better to actually test completion somehow (maybe keep an atomic count of threads that executed and exit when that is correct OR when k_cycle_get_32() is beyond some sanely long timeout).

Don't tune timeouts, basically. It will bite you.